### DISCUSSION TITLE
Should allow useSyncExternalStore for server component

### DISCUSSION THREAD

**Discussion by ysmood** [vercel/next.js#70487] [Help]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/hy4png

### To Reproduce

1. `npm run dev`

### Current vs. Expected behavior

Expect no error, but we will get error:

```txt
You're importing a component that needs useSyncExternalStore.
It only works in a Client Component but none of its parents are marked with "use client",
so they're Server Components by default.
```

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.0.0: Mon Aug 12 20:52:12 PDT 2024; root:xnu-11215.1.10~2/RELEASE_ARM64_T6020
  Available memory (MB): 32768
  Available CPU cores: 12
Binaries:
  Node: 20.12.2
  npm: 9.5.1
  Yarn: 1.22.22
  pnpm: 9.6.0
Relevant Packages:
  next: 14.2.13 // Latest available version is detected (14.2.13).
  eslint-config-next: 14.2.13
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.6.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

create-next-app

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

If you check the doc of React [here](https://react.dev/reference/react/useSyncExternalStore#adding-support-for-server-rendering):

It has a section called Adding support for server rendering, useSyncExternalStore should be able to work for server rendering as long as we provide the third argument getServerSnapshot for it.

**Comment by samcx** [comment]
@ysmood Server Components cannot interface with the client like Client Components. Even if you provide the 3rd argument with `useSyncExternalStore` (getting an initial snapshot from the server), the hook is still interfacing with the client, with the first two arguments. That's why it's only possible to use in a Client Component.

**Reply by ysmood** [reply]
> the hook is still interfacing with the client

I still don't get it, could you provide a link or doc to explain it deeper?

**Reply by ysmood** [reply]
`useSyncExternalStore` works fine in next.js v13 server-side-rendering, why it can't work for server component?

**Reply by samcx** [reply]
@ysmood Maybe the docs isn't so explicit about this, but it is definitely mentioned.

The 3rd param for `useSyncExternalStore` is simply so you can sync data from an external server. But doesn't mean the `useSyncExternalStore` doesn't run on the client.

> If you’re connecting to a third-party data store, you’ll need its data to match between the server and client.

There are multiple mentions of "client" throughout the `useSyncExternalStore` documentation.

> works fine in next.js v13 server-side-rendering

Are you referring to Pages Router?

**Reply by ysmood** [reply]
> The 3rd param for useSyncExternalStore is simply so you can sync data from an external server. But doesn't mean the useSyncExternalStore doesn't run on the client.

Sure it works for "use client", I mean it should also works for server component.

> Are you referring to Pages Router?

Yes, and I can run it fine with offical react server render function like `renderToString`.

Here a test case: [link](https://github.com/ysmood/create-global-state/blob/c30c006991f40d9fd9cfa9160fb81ff8db7bea99/src/index.test.tsx#L100)

**Reply by ysmood** [reply]
Here a more clear example: https://codesandbox.io/p/devbox/sky3nf

If you run it, it will output `<div>ok</div>`. As you can see, it's a pure backend node.js project.

If you comment out the line 8, it will fail with error message `Missing getServerSnapshot, which is required for ' + 'server-rendered content. Will revert to client rendering.`
It means it's designed for server rendering. Why next.js doesn't support it?
